### PR TITLE
✨Add Death module

### PR DESCRIPTION
**Pull request by LQuatre**

### CODE REVIEW

File: src/api/player/server/revive.lua [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+	while not IsScreenFadedOut() do
+		Wait(50)
+	end
+
+	local formattedCoords = {
+		x = API_Maths.round(coords.x, 1),
+		y = API_Maths.round(coords.y, 1),
+		z = API_Maths.round(coords.z, 1)
+	}
```
Comment (CsokiHUN) [inline src/api/player/server/revive.lua:19]: why not use vector3?

File: src/modules/players/server/events/onPlayerDeath.lua [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                        ["KilledByPlayer"] = false
+                    }
+
+                    NCSPlayer.setDeathStatus(_src, NCSisDead, PlayerDeathData)
+                end
+            elseif not IsPedFatallyInjured(PlayerPed) and NCSisDead then
+                waitTime = 0
+                NCSisDead = false
+            end
+        end
```
Comment (CsokiHUN) [inline src/modules/players/server/events/onPlayerDeath.lua:49]:
```suggestion
if (NetworkIsPlayerActive(_src)) then
  local PlayerPed <const> = GetPlayerPed(_src)

  if IsPedFatallyInjured(PlayerPed) and not NCSisDead then
      waitTime = 10
      NCSisDead = true

      local SourceKillerEntity <const>, SourceDeathCause <const> = GetPedSourceOfDeath(PlayerPed)
      local SourceKillerClientId <const> = NetworkGetPlayerIndexFromPed(SourceKillerEntity)
      local PlayerDeathCoords <const> = GetEntityCoords(PlayerPed)

      if SourceKillerEntity ~= _src and SourceKillerClientId and NetworkIsPlayerActive(SourceKillerEntity) then
          local PlayerKillerCoords <const> = GetEntityCoords(SourceKillerEntity)
          local distance <const> = #(PlayerDeathCoords - PlayerKillerCoords)

          local PlayerDeathData = {
              ["PlayerDeathCoords"] = PlayerDeathCoords,
              ["PlayerDeathCause"] = SourceDeathCause,
              ["PlayerDeathTime"] = GetGameTimer(),
              ["KilledByPlayer"] = SourceKillerEntity,
              ["KillerClientId"] = SourceKillerClientId,
              ["KilledByPlayerCoords"] = PlayerKillerCoords,
              ["Distance"] = API_Maths.round(distance, 1)
          }

          NCSPlayer.setDeathStatus(_src, NCSisDead, PlayerDeathData)
      else
          local PlayerDeathData = {
              ["PlayerDeathCoords"] = PlayerDeathCoords,
              ["PlayerDeathCause"] = SourceDeathCause,
              ["PlayerDeathTime"] = GetGameTimer(),
              ["KilledByPlayer"] = false
          }

          NCSPlayer.setDeathStatus(_src, NCSisDead, PlayerDeathData)
      end
  elseif not IsPedFatallyInjured(PlayerPed) and NCSisDead then
      waitTime = 0
      NCSisDead = false
  end
end
```

File: src/class/NCSPlayer/server/functions/setDeathStatus.lua [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    
+    is_dead = 0
+    if DeadStatus then
+        is_dead = 1
+    end
+
+    API_Database:query("UPDATE ncs_players SET is_dead = @is_dead WHERE player_identifier = @player_identifier ", {
+        ["@is_dead"] = is_dead,
+        ["@player_identifier"] = API_Player:getIdentifier(source),
+    })
```
Comment (CsokiHUN) [inline src/class/NCSPlayer/server/functions/setDeathStatus.lua:17]:
```suggestion
API_Database:query("UPDATE ncs_players SET is_dead = @is_dead WHERE player_identifier = @player_identifier ", {
    ["@is_dead"] = (DeadStatus and 1 or 0),
    ["@player_identifier"] = API_Player:getIdentifier(source),
})
```

File: src/api/player/server/revive.lua (hunk 1 above)
Comment (CsokiHUN) [inline src/api/player/server/revive.lua:19]:
```suggestion
local formattedCoords = vector3(
    API_Maths.round(coords.x, 1),
    API_Maths.round(coords.y, 1),
    API_Maths.round(coords.z, 1),
)
```

**Review by CsokiHUN — CHANGES_REQUESTED** [review]
i checked this feature, please check my suggestions

**Review by MoskalykA — CHANGES_REQUESTED** [review]
Please make the changes to a @CsokiHUN

**Comment by LQuatre** [comment]
Thanks, I did this when I was extremely tired,

i will rework all of this, some functions was not used

File: src/api/player/client/spawn.lua [hunk 4]
```diff
@@ -24,12 +24,13 @@ function API_Player:spawn(coords, callback)
 
         local ped <const> = PlayerPedId()
 
-        -- V requires setting coords as well
         SetEntityCoordsNoOffset(ped, position.x, position.y, position.z, false, false, false, true)
         SetEntityHeading(ped, heading)
 
         NetworkResurrectLocalPlayer(position.x, position.y, position.z, heading, true, true, false)
 
+        API_Ped:setInvincible(ped, false)
```
Comment (MoskalykA) [inline src/api/player/client/spawn.lua:32]: Utility ?

File: src/api/player/server/revive.lua [hunk 5]
```diff
@@ -0,0 +1,17 @@
+---revive
+---@param PlayerId number
+---@public
+function API_Player:revive(PlayerId)
```
Comment (MoskalykA) [inline src/api/player/server/revive.lua:4]: Syntax not correct, coords defined?

File: src/class/NCSPlayer/server/functions/minify.lua [hunk 6]
```diff
@@ -9,6 +9,8 @@ function NCSPlayer.minify(self)
         identifier = self.identifier,
         name = self.name,
         role = self.role:minify(),
-        character = self.character:minify()
+        character = self.character:minify(),
+        isDead = self.isDead,
+        deathData = self.deathData or false
```
Comment (MoskalykA) [inline src/class/NCSPlayer/server/functions/minify.lua:14]: Application ?

File: src/class/NCSPlayer/server/functions/setDeathStatus.lua [hunk 7]
```diff
@@ -0,0 +1,13 @@
+---setDeathStatus
+---@param DeadStatus boolean
+---@public
+---@param self NCSPlayer
```
Comment (MoskalykA) [inline src/class/NCSPlayer/server/functions/setDeathStatus.lua:4]:
```suggestion
---@param self NCSPlayer
---@public
```

**Review by MoskalykA — CHANGES_REQUESTED** [review]
I didn't report everything but there are a lot of things to review, please test your code, use a syntax more consistent with the project

File: src/localization/locales/lang_hu.lua [hunk 8]
```diff
@@ -33,4 +33,7 @@ _Literals = {
     KICK_DEFAULT_MESSAGE = "Ki lettél dobva a szerverről.",
     BAN_DEFAULT_MESSAGE = "Ki lettél tiltva a szerverről.",
     UNBAN_DEFAULT_MESSAGE = "A kitiltásod feloldva.",
+
+    ERROR_PLAYER_NOT_DEAD = "Nem vagy még halott.",
```
Comment (CsokiHUN) [inline src/localization/locales/lang_hu.lua:37]:
```suggestion
    ERROR_PLAYER_NOT_DEAD = "Játékos nem halott.",
```

**Review by CsokiHUN — CHANGES_REQUESTED** [review]

**Comment by LQuatre** [comment]
I have to put some pr toi fix some error

**Comment by LQuatre** [comment]
@MoskalykA you can review now

File: src/core/events/client/kill.lua [hunk 9]
```diff
@@ -1,7 +1,5 @@
 NCS:registerNetEvent("ncs_core:killPlayer")
-NCS:handleEvent("ncs_core:killPlayer", function(serverId)
-    local playerId = GetPlayerFromServerId(serverId)
-    local playerPed = GetPlayerPed(playerId)
-    
+NCS:handleEvent("ncs_core:killPlayer", function()
+    local playerPed = GetPlayerPed(-1)
```
Comment (CsokiHUN) [inline src/core/events/client/kill.lua:3]:
```suggestion
    local playerPed <const> = PlayerPedId()
```

**Review by CsokiHUN — CHANGES_REQUESTED** [review]

File: src/class/NCSCharacter/server/_constructor.lua [hunk 10]
```diff
@@ -33,6 +33,14 @@ setmetatable(NCSCharacter, {
         self.metadata = metadata
         self.lastPosition = lastPosition
 
+        API_Database:query("SELECT is_dead FROM ncs_players_characters WHERE character_id = @character_id", {
+            ["@character_id"] = self.id,
+        }, function(rows)
+            local row <const> = rows[1]
+
+            self.isDead = row.is_dead
+        end)
```
Comment (CsokiHUN) [inline src/class/NCSCharacter/server/_constructor.lua:42]:
```suggestion
        API_Database:single("SELECT is_dead FROM ncs_players_characters WHERE character_id = @character_id", {
          ["@character_id"] = self.id,
        }, function(row)
          self.isDead = row.is_dead
        end)
```

File: src/class/NCSCharacter/server/functions/save.lua [hunk 11]
```diff
@@ -0,0 +1,14 @@
+---save
+---@public
+---@param self NCSCharacter
+function NCSCharacter.save(self)
+    API_Database:query("UPDATE ncs_players_characters SET skin = @skin WHERE character_id = @character_id", {
```
Comment (CsokiHUN) [inline src/class/NCSCharacter/server/functions/save.lua:5]:
```suggestion
    API_Database:update("UPDATE ncs_players_characters SET skin = @skin WHERE character_id = @character_id", {
```

File: src/class/NCSPlayer/server/_constructor.lua [hunk 12]
```diff
@@ -37,7 +37,7 @@ setmetatable(NCSPlayer, {
         self.dataLoaded = false
         self.deathData = {}
 
-        API_Database:query("SELECT role_identifier, is_dead FROM ncs_players WHERE player_identifier = @player_identifier", {
+        API_Database:query("SELECT role_identifier FROM ncs_players WHERE player_identifier = @player_identifier", {
```
Comment (CsokiHUN) [inline src/class/NCSPlayer/server/_constructor.lua:40]:
```suggestion
        API_Database:single("SELECT role_identifier FROM ncs_players WHERE player_identifier = @player_identifier", {
```

File: src/class/NCSPlayer/server/_constructor.lua [hunk 13]
```diff
@@ -37,7 +37,7 @@ setmetatable(NCSPlayer, {
...
             ["@player_identifier"] = self.identifier,
         }, function(rows)
             local row <const> = rows[1]
```
Comment (CsokiHUN) [inline src/class/NCSPlayer/server/_constructor.lua:43]:
```suggestion
        }, function(row)
```

File: src/class/NCSPlayer/server/functions/save.lua [hunk 14]
```diff
@@ -4,8 +4,6 @@
 function NCSPlayer.save(self)
     API_Database:query("UPDATE ncs_players SET role_identifier = @role_identifier WHERE player_identifier = @player_identifier", {
```
Comment (CsokiHUN) [inline src/class/NCSPlayer/server/functions/save.lua:5]:
```suggestion
    API_Database:update("UPDATE ncs_players SET role_identifier = @role_identifier WHERE player_identifier = @player_identifier", {
 ```

**Review by CsokiHUN — CHANGES_REQUESTED** [review]

File: src/class/NCSCharacter/server/_constructor.lua (hunk 10 above)
Comment (MoskalykA) [reply src/class/NCSCharacter/server/_constructor.lua:42]: Use async ?

File: src/class/NCSPlayer/server/_constructor.lua (hunk 13 above)
Comment (MoskalykA) [reply src/class/NCSPlayer/server/_constructor.lua:43]: I don't think is good

**Comment by MoskalykA** [comment]
Infinite loading

File: src/class/NCSPlayer/server/_constructor.lua (hunk 13 above)
Comment (CsokiHUN) [reply src/class/NCSPlayer/server/_constructor.lua:43]: see upper change query to single

File: src/core/events/client/revive.lua [hunk 15]
```diff
@@ -0,0 +1,4 @@
+NCS:registerNetEvent("ncs_core:revivePlayer")
+NCS:handleEvent("ncs_core:revivePlayer", function(coords)
+    API_Player:spawn(coords)
+end)
```
Comment (MoskalykA) [inline src/core/events/client/revive.lua:4]:
```suggestion
NCS:registerNetEvent("ncs_core:revivePlayer")
NCS:handleEvent("ncs_core:revivePlayer", function(coords, skin)
    API_Player:spawn(coords)

    CreateThread(function()
        Wait(100)

        MOD_Skin:applySkin(skin)
    end)
end)
```

File: src/modules/commands/server/revivePlayer.lua [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+	local formattedCoords = {
+		position = vector3( coords[1], coords[2], coords[3] ),
+		heading = 90.0
+	}
+	while (not formattedCoords) do 
+		Wait(0)
+	end
+
+    NCSCharacter:setDeathStatus(target.serverId, 0, {})
+    NCS:triggerClientEvent("ncs_core:revivePlayer", target.serverId, formattedCoords)
```
Comment (MoskalykA) [inline src/modules/commands/server/revivePlayer.lua:27]:
```suggestion
    NCS:triggerClientEvent("ncs_core:revivePlayer", target.serverId, formattedCoords, target.character.skin)
```

File: src/modules/players/server/events/onPlayerDeath.lua [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+                else
+                    local playerDeathData = {
+                        ["playerDeathCoords"] = playerDeathCoords,
+                        ["playerDeathCause"] = sourceDeathCause,
+                        ["playerDeathTime"] = GetGameTimer(),
+                        ["killerEntity"] = false
+                    }
+                    NCS:trace(("Player id ^2%s ^7(^2%s^7) died without killer"):format(_src, MOD_Players:get(_src).name), 3)
+                end
+                NCSCharacter:setDeathStatus(_src, isDead, playerDeathData)
```
Comment (MoskalykA) [inline src/modules/players/server/events/onPlayerDeath.lua:32]:
```suggestion
                if sourceKillerEntity ~= _src and sourceKillerEntity  ~= 0 then
                    local playerDeathData = {
                        ["playerDeathCoords"] = playerDeathCoords,
                        ["playerDeathCause"] = sourceDeathCause,
                        ["playerDeathTime"] = GetGameTimer(),
                        ["killerEntity"] = sourceKillerEntity
                    }
                    NCS:trace(("Player id ^2%s ^7(^2%s^7) as been killed by ^2%s ^7(^2%s^7)"):format(_src, MOD_Players:get(_src).name, sourceKillerEntity, MOD_Players:get(sourceKillerEntity).name), 3)
                    NCSCharacter:setDeathStatus(_src, isDead, playerDeathData)
                else
                    local playerDeathData = {
                        ["playerDeathCoords"] = playerDeathCoords,
                        ["playerDeathCause"] = sourceDeathCause,
                        ["playerDeathTime"] = GetGameTimer(),
                        ["killerEntity"] = false
                    }
                    NCS:trace(("Player id ^2%s ^7(^2%s^7) died without killer"):format(_src, MOD_Players:get(_src).name), 3)
                    NCSCharacter:setDeathStatus(_src, isDead, playerDeathData)
                end
```

**Review by MoskalykA — CHANGES_REQUESTED** [review]

File: src/modules/players/server/events/onPlayerDeath.lua [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+NCS:handleEvent("nowInGame", function()
+    local _src <const> = source
+    Wait(5000)
+    if (MOD_Players:exists(_src)) then
+        local player = MOD_Players:get(_src)
+        if (player.character.isDead ~= 0) then
+            NCS:triggerClientEvent("ncs_core:killPlayer", _src)
+        end
+    end
+end)
```
Comment (MoskalykA) [inline src/modules/players/server/events/onPlayerDeath.lua:52]:
```suggestion
```

**Review by MoskalykA — CHANGES_REQUESTED** [review]

**Comment by LQuatre** [comment]
is not functions, is Events*

**Comment by LQuatre** [comment]
PLEASE LAST REVIEW 👍 
I don't have any ideas for additions on this module

**Comment by ach-git** [comment]
I thinks for replace the Wait(5000) when in game you can use src/modules/skin/client/events/onInitializeCharacter.lua:12&18
add: 
```lua
NCS:triggerEvent("initializedCharacter")
```
and replace "nowInGame" by "initializedCharacter"

**Comment by LQuatre** [comment]
> I thinks for replace the Wait(5000) when in game you can use src/modules/skin/client/events/onInitializeCharacter.lua:12&18 add:
> 
> ```lua
> NCS:triggerEvent("initializedCharacter")
> ```
> 
> and replace "nowInGame" by "initializedCharacter"

New event ?

**Review by MoskalykA — APPROVED** [review]

**Pull request merged**